flink-clients/src/test/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrapTest.java

```diff
@@ -233,7 +234,7 @@ public void testApplicationFailsAsSoonAsOneJobFails() throws Throwable {
 
 		final CompletableFuture<Void> applicationFuture = runApplication(dispatcherBuilder, 2);
 
-		assertException(applicationFuture, JobExecutionException.class);
+		assertException(applicationFuture, ApplicationFailedException.class);
```

Comment:
As a side comment concerning the `ApplicationDispatcherBootstrapTest`: It seems a bit strange that I couldn't find a place in the tests where we are calling `ApplicationDispatcherBootstrap.initialize()`. Looking at the interface, this looks like a very important call to do before doing any other operations on the implementations.

Reply:
The problem was that we were passing a `Dispatcher` which is difficult to instantiate and an overkill for such as test. This is the reason why the `runApplicationAndShutdownClusterAsync()` is annotated as `visibleForTesting` and is not private, so that all the tests can call this method.

Reply:
The fact that this worked is actually an indicator that the `ApplicationDispatcherBootstrap` class is following a quite procedural style. Maybe with the change that it now takes a `DispatcherGateway` for which we have easy test implementations one might rethink it.

Reply:
I agree that now we can rethink it, but if I understand the comment correctly, I tend to disagree that this shows that the class follows a procedural style. The only reason that this method was exposed was for testing and it was not accessed by other classes. In that sense, internal state of the class was not "leaking" as even these methods were sealing internal state from outside tampering.